src/backend/vector/packed_simd.rs

```diff
@@ -0,0 +1,252 @@
+            fn bitxor(self, rhs: $ty) -> Self {
+                unsafe { core::arch::x86_64::_mm256_xor_si256(self.0, rhs.0).into_bits() }
+            }
+        }
+
+        impl BitAndAssign for $ty {
+            #[inline]
+            fn bitand_assign(&mut self, rhs: $ty) {
+                *self = *self & rhs;
+            }
+        }
+
+        impl BitXorAssign for $ty {
+            #[inline]
+            fn bitxor_assign(&mut self, rhs: $ty) {
+                *self = *self ^ rhs;
+            }
+        }
+
+        #[allow(dead_code)]
+        impl $ty {
+            #[inline]
+            pub fn shl<const N: i32>(self) -> Self {
+                unsafe { core::arch::x86_64::$shl_intrinsic(self.0, N).into_bits() }
+            }
+
+            #[inline]
+            pub fn shr<const N: i32>(self) -> Self {
+                unsafe { core::arch::x86_64::$shr_intrinsic(self.0, N).into_bits() }
+            }
+
+            #[inline]
+            pub fn extract<const N: i32>(self) -> $lane_ty {
+                unsafe { core::arch::x86_64::$extract_intrinsic(self.0, N) as $lane_ty }
+            }
+        }
+    };
+}
+
+macro_rules! impl_conv {
+    ($src:ident => $($dst:ident),+) => {
+        $(
+            impl IntoBits<$dst> for $src {
+                #[inline]
+                fn into_bits(self) -> $dst {
+                    $dst(self.0)
+                }
+            }
+        )+
+    }
+}
+
+impl_shared!(
+    u64x4,
+    u64,
+    _mm256_add_epi64,
+    _mm256_sub_epi64,
+    _mm256_slli_epi64,
+    _mm256_srli_epi64,
+    _mm256_extract_epi64
+);
+impl_shared!(
+    u32x8,
+    u32,
+    _mm256_add_epi32,
+    _mm256_sub_epi32,
+    _mm256_slli_epi32,
+    _mm256_srli_epi32,
+    _mm256_extract_epi32
+);
+impl_shared!(
+    i32x8,
+    i32,
+    _mm256_add_epi32,
+    _mm256_sub_epi32,
+    _mm256_slli_epi32,
+    _mm256_srli_epi32,
+    _mm256_extract_epi32
+);
+
+impl_conv!(u64x4 => u32x8, i32x8);
+
+#[allow(dead_code)]
+impl u64x4 {
+    #[inline]
+    pub const fn new_const(x0: u64, x1: u64, x2: u64, x3: u64) -> Self {
+        unsafe { Self(core::mem::transmute([x0, x1, x2, x3])) }
+    }
+
+    #[inline]
+    pub const fn splat_const<const N: u64>() -> Self {
+        Self::new_const(N, N, N, N)
+    }
+
+    #[inline]
+    pub fn new(x0: u64, x1: u64, x2: u64, x3: u64) -> Self {
+        unsafe {
+            Self(core::arch::x86_64::_mm256_set_epi64x(
+                x3 as i64, x2 as i64, x1 as i64, x0 as i64,
+            ))
+        }
+    }
+
+    #[inline]
+    pub fn splat(x: u64) -> Self {
+        unsafe { Self(core::arch::x86_64::_mm256_set1_epi64x(x as i64)) }
+    }
+}
+
+#[allow(dead_code)]
+impl u32x8 {
+    #[inline]
+    pub const fn new_const(
+        x0: u32,
+        x1: u32,
+        x2: u32,
+        x3: u32,
+        x4: u32,
+        x5: u32,
+        x6: u32,
+        x7: u32,
+    ) -> Self {
+        unsafe { Self(core::mem::transmute([x0, x1, x2, x3, x4, x5, x6, x7])) }
```

Comment:
Why is this safe?

Reply:
Not resolved. This safety question is separate from the i64 to u64 casting question. I think this behavior relies on the fact that `__mm256i` is a newtype for `(u64, u64, u64, u64)`. Is this representation guaranteed to be the same? Further, is it always the case that a newtype has the same memory layout as its underlying type? And is a tuple of 4 the same as a fixed size array of 4?

Some of these questions might be easy, but for the sake of maintainability (and frankly bc I don't actually know the answer) I think they should be written in a comment here.

Reply:
Oh I apologize @koute I completely missed the comment you added. Resolved